synapse/storage/databases/main/roommember.py

```diff
@@ -620,6 +627,10 @@ def _get_rooms_for_user_with_stream_ordering_txn(
                 room_id, PersistedEventPosition(instance, stream_id)
             )
             for room_id, instance, stream_id in txn
+            if (
+                exclude_for_sync is False
+                or room_id not in self._rooms_excluded_from_sync
+            )
```

Comment:
What's the rationale for doing the filtering here, rather than in the `/sync` code? This is quite a hot cache and so doubling the number of entries isn't particularly ideal

Reply:
This filtering would need to be done for each sync on every user on the server. It makes sense to me to cache it so we don't end up re-doing it on every sync request - and since this path is already cached I figured it would be a good place to do it.

Reply:
That's fair, though in this case we only seem to be calling this function in one place where we immediately filter the results anyway? So we may as well filter the room list there (as its basically free compared to the other functions), and avoid paying the cost of having duplicate caches

Reply:
That would make sense - though we still need to cache the result of the filtering otherwise we end up duplicating `frozenset`s on every sync request.

Reply:
> ... otherwise we end up duplicating `frozenset`s on every sync request.

I think that happens anyway though. Specifically, the only call site of that function simply iterates over the return value and builds a new set:

https://github.com/matrix-org/synapse/blob/f05b984b52993ff96964b1856215f90890c1c775/synapse/handlers/sync.py#L2169-L2198

So adding a `if joined_room.room_id in self.rooms_to_exclude: continue` doesn't make us do any more allocations than before (and the only perf cost is an `if` check, which is cheaper than querying the cached value).

Reply:
Ooooh right I get you now, that makes a lot of sense, thanks!